File: training_data/reflection-tuning/day18_part2_2015.py
def read_input(filename):
    with open(filename, 'r') as f:
        return [list(line.strip()) for line in f]

def count_neighbors(grid, x, y):
    neighbors = 0
    for dx in [-1, 0, 1]:
        for dy in [-1, 0, 1]:
            if dx == 0 and dy == 0:
                continue
            nx, ny = x + dx, y + dy
            if 0 <= nx < len(grid) and 0 <= ny < len(grid[0]):
                neighbors += grid[nx][ny] == '#'
    return neighbors

def update_grid(grid):
    new_grid = [row[:] for row in grid]
    for x in range(len(grid)):
        for y in range(len(grid[0])):
            neighbors = count_neighbors(grid, x, y)
            if grid[x][y] == '#':
                new_grid[x][y] = '#' if neighbors in [2, 3] else '.'
            else:
                new_grid[x][y] = '#' if neighbors == 3 else '.'
    return new_grid

def update_grid_part2(grid):
    new_grid = update_grid(grid)
    n = len(grid) - 1
    new_grid[0][0] = new_grid[0][n] = new_grid[n][0] = new_grid[n][n] = '#'
    return new_grid

def simulate(grid, steps, update_func):
    for _ in range(steps):
        grid = update_func(grid)
    return sum(row.count('#') for row in grid)

def solve_part1(grid):
    return simulate(grid, 100, update_grid)

def solve_part2(grid):
    n = len(grid) - 1
    grid[0][0] = grid[0][n] = grid[n][0] = grid[n][n] = '#'
    return simulate(grid, 100, update_grid_part2)

grid = read_input('input.txt')
print(f"Part 1: {solve_part1(grid)}")
print(f"Part 2: {solve_part2(grid)}")

File: fsharp/day18_part2_2015.fs

open System
open System.IO

let size = 100
let steps = 100
let dirs = [| -1; 0; 1 |]

let countOn (g: bool[,]) x y =
    let mutable c = 0
    for dx in dirs do
        for dy in dirs do
            if dx <> 0 || dy <> 0 then
                let nx = x + dx
                let ny = y + dy
                if nx >= 0 && nx < size && ny >= 0 && ny < size && g.[nx, ny] then
                    c <- c + 1
    c

[<EntryPoint>]
let main _ =
    let lines = File.ReadAllLines("input.txt")
    let grid = Array2D.create size size false
    let newGrid = Array2D.create size size false
    for y = 0 to size-1 do
        let line = lines.[y]
        for x = 0 to size-1 do
            grid.[x, y] <- line.[x] = '#'
    grid.[0,0] <- true
    grid.[0,size-1] <- true
    grid.[size-1,0] <- true
    grid.[size-1,size-1] <- true

    for _ = 1 to steps do
        for x = 0 to size-1 do
            for y = 0 to size-1 do
                let on = countOn grid x y
                newGrid.[x, y] <-
                    if grid.[x, y] then on = 2 || on = 3
                    else on = 3
        newGrid.[0,0] <- true
        newGrid.[0,size-1] <- true
        newGrid.[size-1,0] <- true
        newGrid.[size-1,size-1] <- true
        for x = 0 to size-1 do
            for y = 0 to size-1 do
                grid.[x, y] <- newGrid.[x, y]

    let mutable onCount = 0
    for x = 0 to size-1 do
        for y = 0 to size-1 do
            if grid.[x, y] then onCount <- onCount + 1
    printfn "%d" onCount
    0
